### ISSUE TITLE
Libvirt fails when there are hidden cgroup mount points in `/proc/mounts`

### ISSUE THREAD

**Issue by jhernand** [kubevirt/libvirt#4]
When the the `libvirtd` daemon tries to start a virtual machine it shows the following message in the log:

```
2017-06-30 18:10:10.465+0000: 3585: error : virCgroupMakeGroup:1085 : Failed to create controller net_cls for group: No such file or directory
```

The virtual machine isn't started.

Apparently `libvirt` checks the `/proc/mounts` file to try to find where the cgroup controllers are mounted. In this particular case the `net_cls` controller is listed several times. From the `libvirt` container:

```bash
# grep net_cls /proc/mounts
cgroup /sys/fs/cgroup/net_prio,net_cls cgroup rw,nosuid,nodev,noexec,relatime,net_prio,net_cls 0 0
cgroup /host-sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_prio,net_cls 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_prio,net_cls 0 0
```

Note how `cgroup` appears three times, and the name is in two cases `net_cls,net_prio` and in the other case `net_prio,net_cls`. Note the different order of the cgroup controllers in the mount point.

Checking `/proc/self/mountinfo` I see this:

```bash
# grep net_cls /proc/self/mountinfo
816 810 0:27 /kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-podf1b28fbe_6013_11e7_af2b_525400e6ccec.slice/docker-39b2849efe1a5206cc33bf0b380ee363d3ed74dcea062f287da6efbe7ffa8445.scope /sys/fs/cgroup/net_prio,net_cls rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,net_prio,net_cls
831 825 0:27 / /host-sys/fs/cgroup/net_cls,net_prio rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,net_prio,net_cls
639 633 0:27 / /sys/fs/cgroup/net_cls,net_prio rw,nosuid,nodev,noexec,relatime - cgroup cgroup rw,net_prio,net_cls
```

So apparently Kubernetes (actually Docker) mounted the `net_cls` controller using `net_prio,net_cls` as the file name. That seems wrong, should be `net_cls,net_prio`. That mount point is later hidden by bind mount that we do from `/host-sys` in order to access the host cgroups. The net result is that `/proc/mounts` contains a mount point that isn't actually visible for the container. When libvirt parses `/proc/mounts` it just takes the first line, and fails because the mount point doesn't exist.

I modified libvirt so that it ignores non-existing mount points, with this patch:

```diff
diff --git a/src/util/vircgroup.c b/src/util/vircgroup.c
index 5aa1db5..021a3f2 100644
--- a/src/util/vircgroup.c
+++ b/src/util/vircgroup.c
@@ -393,6 +393,14 @@ virCgroupDetectMountsFromFile(virCgroupPtr group,
         if (STRNEQ(entry.mnt_type, "cgroup"))
             continue;

+        /* Some mount points in the /proc/mounts file may be
+         * hidden by others, and may not actually exist from
+         * the point of the view of the process, so we need
+         * to skip them.
+         */
+       if (!virFileExists(entry.mnt_dir))
+            continue;
+
         for (i = 0; i < VIR_CGROUP_CONTROLLER_LAST; i++) {
             const char *typestr = virCgroupControllerTypeToString(i);
             int typelen = strlen(typestr);
```

That works correctly, but it doesn't pass the libvirt tests: they fail because in the mocked environment none of the mount points exist.

With that change the virtual machine can be started inside kubevirt.

**Comment by jhernand** [comment]
The Kubernetes environment is the following:

```bash
# cat /etc/redhat-release 
CentOS Linux release 7.3.1611 (Core)

# uname -a
Linux kubevirt.local 3.10.0-514.21.2.el7.x86_64 #1 SMP Tue Jun 20 12:24:47 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

# rpm -qa 'kube*' 'docker*' | sort
docker-1.12.6-28.git1398f24.el7.centos.x86_64
docker-client-1.12.6-28.git1398f24.el7.centos.x86_64
docker-common-1.12.6-28.git1398f24.el7.centos.x86_64
kubeadm-1.6.6-0.x86_64
kubectl-1.6.6-0.x86_64
kubelet-1.6.6-0.x86_64
kubernetes-cni-0.5.1-0.x86_64
```

The libvirt environment (the `libivirtd` container inside the `libivirt` pod of KubeVirt) is the following:

```bash
# cat /etc/redhat-release 
Fedora release 25 (Twenty Five)

# rpm -qa 'libvirt*' | sort
libvirt-client-2.2.0-2.fc25.x86_64
libvirt-daemon-2.2.0-2.fc25.x86_64
libvirt-daemon-driver-interface-2.2.0-2.fc25.x86_64
libvirt-daemon-driver-network-2.2.0-2.fc25.x86_64
libvirt-daemon-driver-nodedev-2.2.0-2.fc25.x86_64
libvirt-daemon-driver-nwfilter-2.2.0-2.fc25.x86_64
libvirt-daemon-driver-qemu-2.2.0-2.fc25.x86_64
libvirt-daemon-driver-secret-2.2.0-2.fc25.x86_64
libvirt-daemon-driver-storage-2.2.0-2.fc25.x86_64
libvirt-daemon-kvm-2.2.0-2.fc25.x86_64
libvirt-daemon-qemu-2.2.0-2.fc25.x86_64
libvirt-libs-2.2.0-2.fc25.x86_64
```

All these versions were the newest available at the time I detected the issue.

**Comment by berrange** [comment]
Please report this problem to libvirt upstream

**Comment by rmohr** [comment]
Keeping it open, since it is important for us to track it (and maybe work around it).

**Comment by rmohr** [comment]
@jhernand Thanks for reporting!

**Comment by jhernand** [comment]
@berrange I reported it in this mail to libvirt-list: https://www.redhat.com/archives/libvir-list/2017-July/msg00120.html. Do you prefer a BZ?

**Comment by jhernand** [comment]
The libvirt issue is tracked in this BZ:

  Don't use cgroup mount points from /proc/mounts that are hidden
  https://bugzilla.redhat.com/1468214

**Comment by stu-gott** [comment]
Thanks for catching this @jhernand. For reference, we ran into a similar issue with the networking cgroups (which we were able to work around). This stems from a change in recent versions of systemd where as you note, there is a naming difference between the host system and the container.

**Comment by jhernand** [comment]
The fix for this issue has been merged to the `libvirt` project:

  Avoid hidden cgroup mount points
  http://libvirt.org/git/?p=libvirt.git;a=commitdiff;h=dacd160d7479e0ec2d8a63f102145fd30636a1c8

I opened the following Fedora bug to request a back-port:

  Don't use cgroup mount points from /proc/mounts that are hidden
  https://bugzilla.redhat.com/1470593

**Comment by fabiand** [comment]
Wow - So quickly merged :) Really nice!

**Comment by nertpinx** [comment]
This is fixed in libvirt 3.6.0, we are using muchnewer version, so I propose closing this issue.
